### PR TITLE
Add Soundex

### LINKED ISSUE #30: Add automatic duplicate detection
It'd be lovely to have some sort of way to automatically suggest entries as duplicates, based on similarity of name / title. 

Possbible ways to detect duplicates: 
- Soundex: https://www.php.net/manual/en/function.soundex.php
- Levenshtein https://www.php.net/manual/en/function.levenshtein.php

### PR DESCRIPTION
**Pull request by dan-tang-ssd**
This PR is submitted to fix #30.

It is not yet ready for review, it is submitted for status update.

TODO:
- [x] Add a command to calculate and store soundex result of entity name to soundex column in different entity tables
- [x] In Dimension entry View page, add a modal popup form to show possible duplicates (other dimension entries with same soundex value) as a checklist
- [x] Allow user to select entries and merge (use script that merges the relationships)
- [x] Allow user to mark the entries as not duplicates so they don't appear again as possible duplicates

### CODE REVIEW

**Comment by dan-tang-ssd** [comment]
WIP: In Dimension page, add a button to show potential duplicates (with same soundex value of current record) in modal popup.

---

Screen shots:

image

image

**Comment by dan-tang-ssd** [comment]
I have added (copied) business logic for de-duplication from table bulk action "DeduplicateRecordsAction".

I think it is now ready for review.

---

As Deduplicate is a table row button, I assume user is going to merge the selected potential duplicates to the currently viewing dimension entry. Therefore I skipped the "select one to remain" step.

Some other changes:
1. I splitted "Deduplicate" and "Mark as not duplicate" as two separate table row actions.
2. In Dimension view page, I add the counter for tools relationship.

**Review by dave-mills — CHANGES_REQUESTED** [review]
Functionality looks good. The deduplication and marking entries as not duplicates works as expected. 

### A few UI / UX comments:

- The 'selected' dimension isn't shown anywhere in the modals. I want to know what dimension am I comparing the listed ones to.
- When I see "mark as not duplicate" as a button in the table, it looks like I am about to mark that entry as not a duplicate, rather than review possible duplicates and declare they are not duplicates. 

I am not convinced these work as separate functions - consider possible workflows:
- For a specific dimension, I first want to see what possible duplicates the system has identified. 
- I may want to review the main entry for these possible duplicates (can I open a view page for some of the possible duplicates in a new tab?)
- I then can make a decision for each possible duplicate - is it really a duplicate? Or is it not? 


### Soundex review
Looking at some of the items, I think the soundex is good as a tool, but gives a lot of false positives (check out the screenshot below, which is the identified duplicates for "agricultural runoff" - you don't have to know what that means to notice how widely different the items in this list are). So, while I'm happy keeping the soundex, we definitely need other ways to help us identify duplicates or 'similar-enough' entries. 


CleanShot 2024-03-27 at 14 47 32

**Comment by dave-mills** [comment]
Re: the 2 separated functions - after looking at it a bit more, I agree that we probably shouldn't have both options (merge or mark as not-duplicate) in the same modal popup, at least without a bit more careful UI / UX planning than we have time for. So for now I've removed the 'mark as not duplicate' option from the table. (We could maybe bring it back in on the view page).

Re: Soundex being quite a blunt tool - Looking at the soundex() [page](https://www.php.net/manual/en/function.soundex.php) I saw a reference to metaphone() (under "see also"). I tried that out and it much narrower in identifying duplicates (perhaps too narrow). 

My assessment is that duplicates identified through soundex are "possible" duplicates (but likely not duplicates), while entries with identical metaphone strings are very likely to be duplicates!. 

I have added both lists to the modal popup with an explanation of each, and added icons into the table to show whether an entry has 'likely' duplicates (metaphone) or 'possible' duplicates (soundex).

File: app/Console/Commands/AddMetaphoneToEntities.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+       // $this->addMetaphone(CollectionMethod::class, 'title');
+        $this->addMetaphone(Dimension::class, 'name');
+//        $this->addMetaphone(Scale::class, 'name');
+//        $this->addMetaphone(Tool::class, 'name');
+//        $this->addMetaphone(Framework::class, 'name');
+//        $this->addMetaphone(Unit::class, 'name');
+//        $this->addMetaphone(MetricUser::class, 'name');
+//        $this->addMetaphone(Geography::class, 'name');
+//        $this->addMetaphone(Theme::class, 'name');
```
Comment (dave-mills) [inline app/Console/Commands/AddMetaphoneToEntities.php]: Testing with just dimensions for now

**Comment by dave-mills** [comment]
I think from here we should:

- enable generation of both metaphone and soundex for all the data models;
- merge this into dev and publish it to the staging site along with the other changes.

@dan-tang-ssd can you do this?

**Comment by dan-tang-ssd** [comment]
> I think from here we should:
> 
>     * enable generation of both metaphone and soundex for all the data models;
> 
>     * merge this into dev and publish it to the staging site along with the other changes.
> 
> 
> @dan-tang-ssd can you do this?

Yes, it sounds good.

Should we merge this PR into dev and publish it to staging env first, and then add "Deduplicate" feature to all other entities?

**Comment by dave-mills** [comment]
@dan-tang-ssd - so '12' is a good number for the metaphone from your testing? Sure - lets go for that if you think that gives some better results. 

Once you've enabled the generation of soundex and metaphone for the other entities, this is good to merge in.

**Comment by dan-tang-ssd** [comment]
> @dan-tang-ssd - so '12' is a good number for the metaphone from your testing? Sure - lets go for that if you think that gives some better results.

I did some testing and sampling before, it seems that "12" is a good comprimise. 
Yes, we can finetune it when necessary.

Below please find some sampling for "10", "12" and "15" for quick reference.

---

metaphone "10"

image

---

metaphone "12"

image

---

metaphone "15"

image

**Comment by dan-tang-ssd** [comment]
> Once you've enabled the generation of soundex and metaphone for the other entities, this is good to merge in.

I am going to merge this PR into dev branch now.

I would enable the generation of soundex and metaphone for other entities in future PR that we add "Deplicate" feature to them.

**Comment by dan-tang-ssd** [comment]
In staging env, deployment is done automatically as "Quick Deploy" is enabled.

I checked latest deployment log, error occurred when executing command "npm run build".
"npm run build" was commented in deployment script originally. I have enabled it when comparing deployment script with AEC staging env.

I have just commented command "npm run build" in deployment script, manually triggered "Deploy Now".
Staging env deployment is completed successfully now.

---

In local env, I can replicate the same error when executing command "npm run build".
(P.S. I run command "npm run dev" in local env before, it is my first time to run "npm run build" in local env...)

image

**Pull request merged**